### PR TITLE
Feature/realsense

### LINKED ISSUE #14: Realsense D435/T265 Modules
Create modules hardware abstraction modules for Intel Realsense D435 and T265 modules. [Previous Work](https://github.com/fsherratt/ERL_SmartCities_2019/tree/master/modules/realsense)

### PR DESCRIPTION
**Pull request by fsherratt**

### CODE REVIEW

File: modules/realsense_d435.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+        self._scale = profile.get_device() \
+                        .first_depth_sensor() \
+                        .get_depth_scale()
+                        
+        self._FOV = rs.rs2_fov(self._intrin)
+        self._FOV = np.deg2rad(self._FOV)
+
+    """
+    Initalise conversion matrix for converting the depth frame to a depojected 3D
```
Comment (Neoxas) [inline modules/realsense_d435.py]: Spelling mistake

File: modules/realsense_d435.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
+import pyrealsense2 as rs
+import traceback
+import sys
+import time
+import numpy as np
+
+class rs_d435:
```
Comment (Neoxas) [inline modules/realsense_d435.py:7]: Is there anywhere you can write down the maths for this? Where the formulas come from, general coordinate space etc?

File: modules/realsense_d435.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    Close connection to D435 camera
+    """
+    def close_connection(self):
+        if self._pipe is None:
+            return
+
+        self._pipe.stop()
+        self._pipe = None
+
+        print('rs_t265:D435 Connection Closed')
```
Comment (Neoxas) [inline modules/realsense_d435.py]: Shouldnt this be rs_d435:D435 Connection closed?

File: modules/realsense_t265.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        self._pipe.stop()
+        self._pipe = None
+
+        print('rs_t265:T265 Connection Closed')
+
+    """
+    Retrieve a data from the T265 camera
+    """
+    def get_frame(self) -> tuple:
+        frames = self._pipe.wait_for_frames()
```
Comment (Neoxas) [inline modules/realsense_t265.py]: Is this blocking? If this is blocking, dont we want it runnning in a background thread using the same queue system? Or are we just gonna pump it straight into Rabbit mq?

File: tests/test_realsense_t265.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+from modules import realsense_t265
+
+class TestTemplate(unittest.TestCase):
+    def setUp(self):
+        self.rs = realsense_t265.rs_t265()
+    
+    def tearDown(self):
+        pass
+
+    def test_import_success(self):
```
Comment (Neoxas) [inline tests/test_realsense_t265.py:16]: Rob rasied a good point in that monkeywrenching could be useful here? You could override the pipe method to a class of your chosing and use that to feed data in? Then you could use that to ensure your private functions work as expected.

File: tests/test_realsense_d435.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+from modules import realsense_d435
+
+class TestTemplate(unittest.TestCase):
+    def setUp(self):
+        self.rs = realsense_d435.rs_d435()
+    
+    def tearDown(self):
+        pass
+
+    def test_import_success(self):
```
Comment (Neoxas) [inline tests/test_realsense_d435.py:16]: Rob rasied a good point in that monkeywrenching could be useful here? You could override the pipe method to a class of your chosing and use that to feed data in? Then you could use that to ensure your private functions work as expected.

**Review by Neoxas — CHANGES_REQUESTED** [review]
Minor items. Probalby worth looking at if monkeywrenching is useful for overwriding pipes etc. Also would be worth adding unit tests for the private matrix conversion functions to ensure the operations we expect happen. Plus a design document for the maths in use would be good.

File: modules/realsense_t265.py (hunk 4 above)
Comment (fsherratt) [reply modules/realsense_t265.py]: It's blocking, was assuming this was just going to be a single thread producer pushing straight to rabbit MQ

File: modules/realsense_d435.py (hunk 1 above)
Comment (fsherratt) [reply modules/realsense_d435.py]: Fixed in 8ffb4b6

File: modules/realsense_d435.py (hunk 3 above)
Comment (fsherratt) [reply modules/realsense_d435.py]: Fixed in 8ffb4b6

File: modules/realsense_d435.py (hunk 2 above)
Comment (fsherratt) [reply modules/realsense_d435.py:7]: I guess I should put a markdown document together about both cameras. It's is a little complicated to explain though. This is where the [deprojection](https://github.com/IntelRealSense/librealsense/blob/7148f9ae1d78b5d44bee4fc578bf0b8fb9a220c5/wrappers/python/examples/box_dimensioner_multicam/helper_functions.py#L121) comes from

Comment (fsherratt) [reply modules/realsense_d435.py:7]: First stab at documentation is here 4cad3967ded27b7fe7eef6c871b2c15e4721b04c

File: tests/test_realsense_d435.py (hunk 6 above)
Comment (fsherratt) [reply tests/test_realsense_d435.py:16]: Had a stab at some unit tests c6cd50b. Not convinced by all of them

File: tests/test_realsense_t265.py (hunk 5 above)
Comment (fsherratt) [reply tests/test_realsense_t265.py:16]: 03eac75 first attempt

**Comment by fsherratt** [comment]
Unit tests have now been written for all functionality and ready for review

File: modules/realsense_d435.py (hunk 2 above)
Comment (Neoxas) [reply modules/realsense_d435.py:7]: Happy with that

**Review by Neoxas — DISMISSED** [review]
Happy with all these changes

**Review by Neoxas — APPROVED** [review]

**Pull request merged**